language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -302,18 +326,104 @@ unfoldFixOf pat arg k = do
+          -- ^ annoying: we have to normalise during subsitution, but
+          -- then we need a non-normalised type to perform subsequent
+          -- subsitutions on.
+    ty' <- foldM f (void ty) mapping
+    normalizeTypeM ty'
+
+-- | Work along the type parameters of a builtin application, building
+-- a mapping from type parameters to the types they're being
+-- instantiated at.
+checkBuiltinTypeArgs
+    :: ann
+    -> BuiltinName
+    -> [TyName]                -- type parameters
+    -> [Type TyName uni ann]   -- type instantations
+    -> TypeCheckM uni ann (TypeMapping uni)
+checkBuiltinTypeArgs = check [] where
+    check mapping ann bn (tyname:tynames) (tyarg:tyargs) = do
+       checkKindM  (typeAnn tyarg) tyarg (Type ()) -- NB: All of the type variables in TypeSchemes have kind *
+       tyarg' <- normalizeType (void tyarg)
+       check ((tyname,tyarg'):mapping) ann bn tynames tyargs
+    check mapping _ _ [] [] = pure mapping
+    check _ ann bn _ [] = throwError $ BuiltinTypeErrorE ann (BuiltinUnderInstantiated bn)
+    check _ ann bn [] _ = throwError $ BuiltinTypeErrorE ann (BuiltinOverInstantiated  bn)
+
+-- | Given a mapping from type variables to types, we work our way
+-- along the list of term arguments in an ApplyBuiltin, checking that
+-- the types of the arguments match the types given in the signature
+-- after instatiation according to the mapping.
+checkBuiltinTermArgs
+    :: (GShow uni, GEq uni, DefaultUni <: uni)
+    => TypeMapping uni
+    -> ann
+    -> BuiltinName
+    -> [Type TyName uni ()]         -- expected argument types
+    -> [Value TyName Name uni ann]  -- actual arguments
+    -> TypeCheckM uni ann ()
+checkBuiltinTermArgs mapping ann bn (ty:tys) (arg:args) =  do
+    ty' <- substMapping mapping ty  -- normalised
+    checkTypeM (termAnn arg) arg ty'
+    checkBuiltinTermArgs mapping ann bn tys args
+checkBuiltinTermArgs _ _ _ [] [] = pure ()
+checkBuiltinTermArgs _ ann bn _ [] = throwError $ BuiltinTypeErrorE ann (BuiltinUnderApplied bn)
+checkBuiltinTermArgs _ ann bn [] _ = throwError $ BuiltinTypeErrorE ann (BuiltinOverApplied  bn)
+
+-- | Check an entire ApplyBuiltin term: make sure that it has the right number of
+-- type and term arguments and that the term argments have the right type.  Return
+-- the return type, correctly instantiated.
+checkBuiltinAppl
+    :: (GShow uni, GEq uni, DefaultUni <: uni)
+    => ann
+    -> BuiltinName
+    -> [TyName]                    -- Type parameters of builtin
+    -> [Type TyName uni ann]       -- Actual type instantations
+    -> [Type TyName uni ()]        -- Types of arguments
+    -> [Value TyName Name uni ann] -- Actual arguments
+    -> Type TyName uni ()          -- Expected return type
+    -> TypeCheckM uni ann (Normalized (Type TyName uni ()))
+checkBuiltinAppl ann bn typeVars typeArgs argTypes args rtype = do
+  mapping <- checkBuiltinTypeArgs ann bn typeVars typeArgs
+  checkBuiltinTermArgs mapping ann bn argTypes args
```

Comment:
We could remove the `mapping` argument by discharging the `mapping` into `argTypes` via `map (substMapping mapping)`. Then `checkBuiltinTermArgs` doesn't need to know anything about mappings and can just directly type check arguments, which makes the code less tightly coupled.